### PR TITLE
Implementation required to enable Forwarding if it is already disabled

### PR DESCRIPTION
**Pull request by arghosh93**
This PR is to remove any iptables rules which were created based on config.Gateway.DisableForwarding set to true and later it was changed to false.

This PR also ensures forwarding sysctl parameters are set for all IPv4 and IPv6 interfaces by either enabling or disabling it based on config.Gateway.DisableForwarding

**- What this PR does and why is it needed**

**- Special notes for reviewers**

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

**Comment by arghosh93** [comment]
I have added a function setGlobalForwardingRule() to enable or disable forwarding for all IPv4 and IPv6 interfaces based on config.Gateway.DisableForwarding. 
Currently we enable forwarding sysctl parameters for Global ipForwarding mode and we do set it while starting[1] ovnkube-node POD. But we dont rollback forwarding sysctl parameters to default value if we change ipForwarding mode from Global to Restricted but I think we should do that. Instead of taking care of this in here we can take care of this in ovnkube-node startup script[1]. I am open to suggestions on this point.  

[1] - https://github.com/openshift/cluster-network-operator/blob/master/bindata/network/ovn-kubernetes/common/008-script-lib.yaml#L568

**Comment by tssurya** [comment]
ack let's talk about this a bit in our next meeting

File: go-controller/pkg/node/default_node_network_controller.go
Comment (arghosh93) [reply go-controller/pkg/node/default_node_network_controller.go:1489]: `so IIUC we don't need to do https://github.com/openshift/cluster-network-operator/commit/e77c0f67cb9250bf94d2b1a667d7afadda33add8 with this specific fix?`
--> Yes, we dont have to set sysctl parameters before starting ovnkube in that case.

File: go-controller/pkg/node/default_node_network_controller.go
Comment (trozet) [inline go-controller/pkg/node/default_node_network_controller.go:1488]: So OVN-Kubernetes should not be setting global ip forwarding for anything. That is the role of the system admin of the node. It can cause undesired consequences if we globally enable forwarding for all interfaces on a server. So the stance of OVN-Kubernetes is we will set forwarding for the interfaces we manage.

The config is specific about this:
```
	&cli.BoolFlag{
		Name:        "disable-forwarding",
		Usage:       "Disable forwarding on OVNK controlled interfaces.",
		Destination: &cliConfig.Gateway.DisableForwarding,
	},
```

So I don't think you should set any global forwarding values. I also don't think you need to mess with the per interface forwarding regardless of what the forwarding mode is set to. It is required to allow some forwarding on OVNK managed interfaces for OVNK to function. Therefore the sysctl enablement is required no matter what. However, we can remove the restricted iptables rules when the flag is switched.

**Review by trozet — CHANGES_REQUESTED** [review]

File: go-controller/pkg/node/default_node_network_controller.go
Comment (arghosh93) [reply go-controller/pkg/node/default_node_network_controller.go:1488]: @trozet While starting ovnkube-node we set[1] net.ipv4.ip_forward sysctl parameter to '1' when ipForwarding mode is set to 'Global' which in turn enables forwarding for all IPv4 interfaces. However we dont disable forwarding by setting  net.ipv4.ip_forward sysctl parameter to '0' when ipForwarding mode is switched to 'Restricted' from 'Global'. In newly created >4.14 cluster this bug does not impact as default ipForwarding mode is Restricted and net.ipv4.ip_forward sysctl parameter gets set to '0' by default. 

I do agree that we should not set IPv4 IP forwarding for all interfaces at ovn-kubernetes level. If ipForwarding mode is set to 'Restricted' then we can set net.ipv4.ip_forward sysctl parameter to '0' while starting[2] ovnkube-node itself. I can open a CNO PR additionally if you think that is ideal.

Addition and deletion of iptables rules can be taken care at ovn-kubernetes level as you agreed.

[1] - https://github.com/openshift/cluster-network-operator/blob/master/bindata/network/ovn-kubernetes/common/008-script-lib.yaml#L568
[2] - https://github.com/openshift/cluster-network-operator/blob/master/bindata/network/ovn-kubernetes/common/008-script-lib.yaml#L570

Comment (trozet) [reply go-controller/pkg/node/default_node_network_controller.go:1488]: Right. That is an OpenShift specific thing that should be done in CNO where you linked.

Comment (arghosh93) [reply go-controller/pkg/node/default_node_network_controller.go]: I have raised a CNO PR as per agreement. https://github.com/openshift/cluster-network-operator/pull/2385
Thanks Tim & Surya :)

File: go-controller/pkg/node/gateway_iptables.go
Comment (tssurya) [inline go-controller/pkg/node/gateway_iptables.go:388]: do we need the extra abstraction here? can't we just call `nodeipt.DelRules(getGatewayForwardRules(cidrs))` straight from here? its straightforward enough for readability that I don't see the need for abstracting

File: go-controller/pkg/node/gateway_localnet.go
Comment (tssurya) [inline go-controller/pkg/node/gateway_localnet.go:86]: could you please add a unit test for this; you can see samples: https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/node/gateway_localnet_linux_test.go

1. do one for addition with config=true
2. then use the same test to change config=false and then re-run to see if we are deleting the rules

File: go-controller/pkg/node/gateway_shared_intf.go
Comment (tssurya) [inline go-controller/pkg/node/gateway_shared_intf.go:1857]: nit: 
```suggestion
			return nil, fmt.Errorf("failed to delete drop rules in forwarding table for bridge %s: err %v", gwBridge.bridgeName, err)
```
or something
so that we can distinguish which func failed both functions are returning same snippet; I'd suggest the same change for the add piece above and same for the above code snippets as well

**Review by tssurya — CHANGES_REQUESTED** [review]
main logic looks good, please add tests for this and then I can approve

I won't force you but ... if you have time are you interested in writing up a short documentation on "disable-forwarding" config option under https://ovn-kubernetes.io/developer-guide/configuration/ please? We never had the culture of documenting anything before so I know it might seem unfair to ask you to do this now so feel free to ignore my suggestion but if you do have time.. then all I need is a 
subtitle saying "Disable Forwarding Config" => then write details on what this does and add all the sample iptables rules it adds and how one can set it to false to avoid these rules etc.. (PS: Don't add any OCP product info BUT deff mention the use case for this config option and then instead of using OCP/CNO etc use operator or admin or whatever generic terms suits here)

File: go-controller/pkg/node/gateway_iptables.go
Comment (arghosh93) [reply go-controller/pkg/node/gateway_iptables.go:388]: I have done this to keep consistency with the initExternalBridgeServiceForwardingRules() and initExternalBridgeDropForwardingRules() function. init functions uses this abstraction mostly due to use of insertIptRules() in one and appendIptRules() in the other init function.

File: go-controller/pkg/node/gateway_localnet.go
Comment (arghosh93) [reply go-controller/pkg/node/gateway_localnet.go:86]: Ack

File: go-controller/pkg/node/gateway_shared_intf.go
Comment (arghosh93) [reply go-controller/pkg/node/gateway_shared_intf.go:1857]: Ack

File: go-controller/pkg/node/gateway_iptables.go
Comment (tssurya) [reply go-controller/pkg/node/gateway_iptables.go:388]: @arghosh93 ^

**Review by tssurya — APPROVED** [review]

**Comment by tssurya** [comment]
docs will be done in new PR

**Comment by tssurya** [comment]
DCO seems confused, not sure why.

**Comment by arghosh93** [comment]
Unit test passed in local environment. Unrelated flake from ovn-kubernetes/go-controller/pkg/ovn/egressip_test.go:1981

/retest-required

File: docs/getting-started/configuration.md
Comment (tssurya) [inline docs/getting-started/configuration.md:1]: lgtm! can you please change this also here : https://github.com/ovn-org/ovn-kubernetes/blob/47fcdd2607be4e4b9cc198a0d6fca385bf4030ed/mkdocs.yml#L89 so that it appears on our website?
use https://ovn-kubernetes.io/developer-guide/documentation/#how-to-test-your-documentation-changes to test your changes

Comment (arghosh93) [reply docs/getting-started/configuration.md:1]: Done

**Review by tssurya — APPROVED** [review]
Will merge after https://github.com/dcoapp/app/issues/211 get's sorted out

**Comment by tssurya** [comment]
CI is green, only the DCO lane is blocked but I have manually verified the signoff is correct. tysm @arghosh93 !!

**Pull request merged**
